build/plugins/attributes.html

```diff
@@ -0,0 +1,132 @@
+<!-- attributes plugin -->
+
+<script>
+    (function() {
+        // /////////////////////////
+        // DESCRIPTION
+        // /////////////////////////
+
+        // This Manubot plugin allows arbitrary HTML attributes to be attached
+        // to (almost) any element. Place an HTML comment inside or next to the
+        // desired element in the format <!-- attribute="value" -->
+
+        // /////////////////////////
+        // OPTIONS
+        // /////////////////////////
+
+        // plugin name prefix for url parameters
+        const pluginName = 'attributes';
+
+        // default plugin options
+        const options = {
+            // whether plugin is on or not
+            enabled: 'true'
+        };
+
+        // change options above, or override with url parameter, eg:
+        // 'manuscript.html?pluginName-enabled=false'
+
+        // /////////////////////////
+        // SCRIPT
+        // /////////////////////////
+
+        // start script
+        function start() {
+            // get list of comments in document
+            const comments = findComments();
+
+            for(const comment of comments) {
+                // if comment doesn't have valid parent element
+                if (!comment.parentElement)
+                    continue;
+
+                // get text content of comment
+                let text = comment.nodeValue.trim();
+
+                // regex's for finding attribute/value pairs in the format of
+                // attribute="value" or attribute='value
+                const regex2 = /([a-zA-Z\-]+)?=\"(.+?)\"/;
+                const regex1 = /([a-zA-Z\-]+)?=\'(.+?)\'/;
```

Comment:
I see `-` is allowed in attribute but not `_`. Do HTML attributes not use underscores?

Does this only match from the start of the comment string? For example what happens if:

```markdown
<!-- The previous sentence is needed to describe when method="funk" -->
```

Probably should make syntax isn't detected by accident (although if it were, the effects would probably be minor?).

The initial comment uses `<!-- colspan: 2 -->`, but that syntax is different than `<!-- colspan="2" -->`

Reply:
Standard HTML attributes only have letters (`colspan` `style` `class`) or dashes (`data-somevar` `stroke-width`). Custom attributes can have more, but custom attributes won't do anything without associated javascript, so at that point the person has to be a javascript developer anyway.

See this for more info: https://stackoverflow.com/questions/925994/what-characters-are-allowed-in-an-html-attribute-name

Also no, it doesn't need to be at the start of the string. That is intentional. And it matches multiple, so you can put more than one attribute (it'd be frustrating to the user for us to make this plugin but then only allow one attribute). Also, I think other notes should be allowed right next to the attributes.

Therefore I think the behavior is correct as it is. `<!-- this is some extra styling style="color: red" and here's a colspan colspan="2" yay -->`

Reply:
I agree that defining multiple attributes in a single comment is important.

I don't like however that the plugin scans every comment and extracts things where the user may not actually be intending to set attributes. In your example, the user does intent to set attributes. In my example, the user didn't. Therefore, would it make sense to require something at the beginning of the comment like:

```
<!-- htmlattrs colspan="2" style="color: red" -->
```

Then you could strip leading whitespace from all comments and ignore any comment that doesn't begin with `htmlattrs` (or whatever keyword we choose).

Finally, it seems that the `colspan="2" style="color: red"` definitions are adopting a standard syntax for HTML? If so, would it be possible to use the builtin infrastructure to parse this text? This way it could deal with more complex cases (possibly with escaping etc) and would work exactly as if this text had been actual HTML.

One final consideration is how the system will work for non-English users. It seems like HTML tags are always in English, such that this is not a problem?

Reply:
I don't know of any existing infrastructure to do this. You're not really intended to do anything like this in normal practice. There is `setAttribute` and `getAttribute` for setting/getting attributes, but that is javascript.

As far as I know, yes, the attribute names are always in english. Attributes that accept strings, like `title` or `alt`, I believe can accept any characters/languages.

Regarding `htmlattrs`, we can talk more about how you'd like to set this up such that users only do it intentionally.

Reply:
Is this the right functionality? https://api.jquery.com/jQuery.parseHTML/

Reply:
Not really, no. That jQuery function parses a full html string, like `<span class="button">text</span>`, and returns a DOM node that can be inserted into the document.

In our case, we already have the DOM node (the node that the comment is attached to/under) and just need to add attributes to it.

What you're looking for (I think) is an existing function to read a string and extract out valid HTML attribute/value pairs. I don't believe that exists.

As for your second links, they do give me an idea:

I can duplicate the parent node, find the end of the opening tag, and insert the comment string into it like `<span this text was originally in a comment colspan="2">`, read that duplicate's attributes with `getAttribute`, and apply the same values to the real parent node.

But I think this is more error prone, more verbose, still requires some parsing (to find the starting tag and add text to it), and still not a standard procedure.

Reply:
If we dissallow non-attribute comments after colspan, would this be straightword:

From here:

```html
<!-- htmlattrs colspan="2" style="color: red" -->
```

To here:

```
<span colspan="2" style="color: red" ></span>
```

To parsing the attributes and extracting them?

Reply:
That particular change doesn't help or hurt.

I can rewrite the plugin using the idea in my previous comment, and you can see how you like the readability of it compared to the current method. The new method is less predictable, to me. I can't be sure there won't be unforeseen odd edge cases that cause trouble down the road.

Reply:
After experimentation, I've concluded that the "new" method can't actually work because cloning the node adequately is not possible. You can verify this for yourself by running these snippets in the console:

Try to clone object by first creating container then setting `innerHTML` with new desired attribute string from comment
```javascript
container = document.createElement("div");
container.innerHTML = "<td newAttribute='fromComment'></td>";
console.log(div);
```
The resulting `<div>` node contains nothing because a `<td>` (table cell) is not valid unless it's a child of a `<tr>` (row) and `<table>`.

This problem also occurs using `DOMParser`. There are certain combinations and fragments of HTML that are not allowed, and it would be too difficult to capture them all and handle the exceptions. It would almost be like rewriting an HTML display engine.

Also, for the situations where the above approach (and the several variants that I also tried) DOES work, it still doesn't seem to "purify" the input in any useful way. It still returns all of the attributes in a list, even if they're not valid HTML attributes. It seems to just split the string by spaces and `=`'s. Try out this code:

```javascript
div = document.createElement("div");
div.innerHTML = "<span text: text colspace='2'>text</span>";
console.log(div.firstChild.attributes);
```

Therefore I don't think this is better in any way than the current approach.